library/x509_create.c

```diff
@@ -86,73 +92,214 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
+                    {
+                        ret = MBEDTLS_ERR_X509_INVALID_NAME;
+                        goto exit;
+                    }
+                }
             }
-        }
-        else if( !in_tag && ( *c == ',' || c == end ) )
-        {
-            if( mbedtls_asn1_store_named_data( head, oid, strlen( oid ),
-                                       (unsigned char *) data,
-                                       d - data ) == NULL )
-            {
-                return( MBEDTLS_ERR_X509_ALLOC_FAILED );
-            }
-
-            while( c < end && *(c + 1) == ' ' )
-                c++;
-
-            s = c + 1;
-            in_tag = 1;
-        }
-
-        if( !in_tag && s != c + 1 )
-        {
-            *(d++) = *c;
+            break;
 
-            if( d - data == MBEDTLS_X509_MAX_DN_NAME_SIZE )
+            /*
+             * Currently parsing an unquoted value.
+             */
+            case dn_in_unquoted_value:
             {
-                ret = MBEDTLS_ERR_X509_INVALID_NAME;
-                goto exit;
+                if( *cur == ',' || cur == end )
+                {
+                    if( d - data == 0 )
+                    {
+                        ret = MBEDTLS_ERR_X509_INVALID_NAME;
+                        goto exit;
+                    }
+
+                    if( mbedtls_asn1_store_named_data( head, oid, strlen( oid ),
+                                                       (unsigned char *) data,
+                                                       d - data ) == NULL )
+                    {
+                        ret = MBEDTLS_ERR_X509_ALLOC_FAILED;
+                        goto exit;
+                    }
+                    /*
+                     * Skip white spaces
+                     */
+                    while( cur < end && *(cur + 1) == ' ' )
+                        cur++;
+
+                    start = cur + 1;
+                    dn_cur_state = dn_in_tag;
+                    tag_len = 0;
+                }
+                else
+                {
+                    if( *cur == '\\' && cur != end )
+                    {
+                        cur++;
+
+                        ret = x509_dn_check_special_chars( &cur, end );
+                        if( ret != 0 )
+                        {
+                            ret = MBEDTLS_ERR_X509_INVALID_NAME;
+                            goto exit;
+                        }
+                    }
+                    /*
+                     * Fill the current data pointer with the unescaped character.
+                     */
+                    *(d++) = *cur;
+
+                    if( d - data == MBEDTLS_X509_MAX_DN_NAME_SIZE )
+                    {
+                        ret = MBEDTLS_ERR_X509_INVALID_NAME;
+                        goto exit;
+                    }
+                }
             }
         }
 
-        c++;
+        cur++;
+    }
+
+    if( dn_cur_state != dn_in_tag )
```

Comment:
With this, we could accept a DN string like `A=B C`, ignoring the `C`. We might fix this by requiring `start == end` I think.

Reply:
don't think it will work. `start` will always be equal to `end + 1` when we finish the operation.
I think that the whole whitespace handling requires refactoring. How about adding 4 more states:
* `dn_before_tag'  - will be initialization state, will allow whitespace, and transition to `dn_in_tag` once characters arrive. otherwise - error.
*  `dn_after_tag` - will be transitioned from `dn_in_tag` if there is white space. Allows only whitespace and `=`
* `dn_before_value` - same as `dn_before_tag` but for value.
* `dn_after_value` - same as `dn_after_tag` but for value.

Reply:
@RonEld I fully agree that a proper solution to this needs enhancing the state machine, and I'd be happy if you'd take the effort to implement it. In your proposal, I think `dn_after_value` needs no separation from `dn_before_tag`, or does it?

Reply:
`dn_after_value` should allow whitespaces, and transition to `dn_before_tag` once a comma arives.
`dn_before_tag` shouold allow whitespaces and transition to `dn_in_tag` once valid characters arrive.

I'll add the changes